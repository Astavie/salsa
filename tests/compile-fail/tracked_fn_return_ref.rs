use salsa::Database as Db;

#[salsa::input]
struct MyInput {
    #[return_ref]
    text: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct ContainsRef<'db> {
    text: &'db str,
}

#[salsa::tracked]
fn tracked_fn_return_ref<'db>(db: &'db dyn Db, input: MyInput) -> &'db str {
    input.text(db)
}

#[salsa::tracked]
fn tracked_fn_return_struct_containing_ref<'db>(
    db: &'db dyn Db,
    input: MyInput,
) -> ContainsRef<'db> {
    ContainsRef {
        text: input.text(db),
    }
}

fn main() {}
