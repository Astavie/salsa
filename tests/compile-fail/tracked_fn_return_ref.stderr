warning: unused import: `salsa::Update`
 --> tests/compile-fail/tracked_fn_return_ref.rs:2:5
  |
2 | use salsa::Update;
  |     ^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error: lifetime may not live long enough
  --> tests/compile-fail/tracked_fn_return_ref.rs:15:1
   |
15 | #[salsa::tracked]
   | ^^^^^^^^^^^^^^^^^
   | |
   | lifetime `'db` defined here
   | requires that `'db` must outlive `'static`
   |
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error: lifetime may not live long enough
  --> tests/compile-fail/tracked_fn_return_ref.rs:20:1
   |
20 | #[salsa::tracked]
   | ^^^^^^^^^^^^^^^^^
   | |
   | lifetime `'db` defined here
   | requires that `'db` must outlive `'static`
   |
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)
